### PR TITLE
Make the RMT library non-blocking.

### PR DESCRIPTION
**Pull request by floitsch**

### CODE REVIEW

File: lib/rmt.toit [hunk 1]
```diff
@@ -196,23 +214,24 @@ class Channel:
...
   static DEFAULT_OUT_CARRIER_LEVEL ::= 1
   static DEFAULT_OUT_CARRIER_DUTY_PERCENT ::= 33
   static DEFAULT_OUT_IDLE_LEVEL ::= null
-  static DEFAULT_READ_TIMEOUT_MS ::= 100
 
   pin       /gpio.Pin
   resource_ /ByteArray? := ?
 
-  rx_buffer_size_/int? := null
+  static NOT_CONFIGURED_ ::= 0
```
Comment (kasperl) [inline lib/rmt.toit:228]: I often prefer sharing a prefix for these constants (CONFIGURED_xxx), but I'm not sure that would be great here.

File: lib/rmt.toit [hunk 2]
```diff
@@ -356,59 +412,117 @@ class Channel:
...
-  read max_bytes/int --timeout_ms/int=DEFAULT_READ_TIMEOUT_MS -> Signals:
+  read --stop_reading/bool?=null -> Signals:
     if not is_input: throw "INVALID_STATE"
-    if max_bytes > buffer_size: throw "maximum returned buffer size greater than allocated buffer size"
-    bytes := rmt_receive_ resource_ max_bytes timeout_ms
-    return Signals.from_bytes bytes
+
+    was_reading := is_reading_
+    if stop_reading == null: stop_reading = not was_reading
+
```
Comment (kasperl) [inline lib/rmt.toit:506]: I'd move the newline onto the other side of the 'if not was_reading' line.

File: src/resources/rmt_esp32.cc [hunk 3]
```diff
@@ -256,12 +264,42 @@ PRIMITIVE(transmit) {
...
+    address = bytes.address();
+    keep_alive = external_copy;
+  }
+
   rmt_channel_t channel = resource->channel();
-  const rmt_item32_t* items = reinterpret_cast<const rmt_item32_t*>(items_bytes.address());
-  esp_err_t err = rmt_write_items(channel, items, items_bytes.length() / 4, true);
-  if ( err != ESP_OK) return Primitive::os_error(err, process);
+  const rmt_item32_t* items = reinterpret_cast<const rmt_item32_t*>(address);
+  bool wait_until_done;
```
Comment (kasperl) [inline src/resources/rmt_esp32.cc:289]: bool wait_until_done;  // Local for naming argument to rmt_write_items.

File: src/resources/rmt_esp32.cc [hunk 4]
```diff
@@ -272,117 +310,73 @@ static void flush_buffer(RingbufHandle_t rb) {
...
+  // If resize is true, then we expect that the output array was allocated in prepare_receive.
+  // As such it is an external byte array that we can resize.
+  ARGS(RMTResource, resource, ByteArray, output, bool, resize)
+
+  RingbufHandle_t rb = null;
+  esp_err_t err = rmt_get_ringbuf_handle(resource->channel(), &rb);
+  if (err != ESP_OK) return Primitive::os_error(err, process);
 
+  size_t received_length;
+  auto received_bytes = xRingbufferReceive(rb, &received_length, 1);
```
Comment (kasperl) [inline src/resources/rmt_esp32.cc:354]: 1 -> 0 ?

File: tests/esp32/rmt.toit [hunk 5]
```diff
@@ -6,58 +6,25 @@
...
-  reset := gpio.Pin TESTER_RESET_OUT_PIN --output
-  reset.set 0
-  sleep --ms=100
-  reset.set 1
-  reset.close
+RMT_PIN_1 ::= 18
+RMT_PIN_2 ::= 19
+RMT_PIN_3 ::= 15
 
+// Because of the resistors we use the reading isn't fully precise.
```
Comment (kasperl) [inline tests/esp32/rmt.toit:25]:
```suggestion
// Because of the resistors we use, the reading isn't fully precise.
```

**Review by kasperl — APPROVED** [review]
Looks good to me!

File: lib/rmt.toit (hunk 1 above)
Comment (floitsch) [reply lib/rmt.toit]: CONFIGURED_NONE_
done.

File: lib/rmt.toit (hunk 2 above)
Comment (floitsch) [reply lib/rmt.toit]: done.

File: src/resources/rmt_esp32.cc (hunk 3 above)
Comment (floitsch) [reply src/resources/rmt_esp32.cc]: done.

File: src/resources/rmt_esp32.cc (hunk 4 above)
Comment (floitsch) [reply src/resources/rmt_esp32.cc]: done.

File: tests/esp32/rmt.toit (hunk 5 above)
Comment (floitsch) [reply tests/esp32/rmt.toit]: done.

**Pull request merged**